ee/middleware/upstreamoauth/provider.go

```diff
@@ -0,0 +1,300 @@
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	"golang.org/x/oauth2"
+	oauth2clientcredentials "golang.org/x/oauth2/clientcredentials"
+
+	"github.com/TykTechnologies/tyk/apidef"
+	"github.com/TykTechnologies/tyk/internal/crypto"
+	"github.com/TykTechnologies/tyk/internal/model"
+)
+
+// Provider implements upstream auth provider.
+type Provider struct {
+	// Logger is the logger to be used.
+	Logger *logrus.Entry
+	// HeaderName is the header name to be used to fill upstream auth with.
+	HeaderName string
+	// AuthValue is the value of auth header.
+	AuthValue string
+}
+
+// Fill sets the request's HeaderName with AuthValue
+func (u Provider) Fill(r *http.Request) {
+	if r.Header.Get(u.HeaderName) != "" {
+		u.Logger.WithFields(logrus.Fields{
+			"header": u.HeaderName,
+		}).Info("Authorization header conflict detected: Client header overwritten by Gateway upstream authentication header.")
+	}
+	r.Header.Set(u.HeaderName, u.AuthValue)
+}
+
+type OAuthHeaderProvider interface {
+	// getOAuthToken returns the OAuth token for the request.
+	getOAuthToken(r *http.Request, OAuthSpec *Middleware) (string, error)
+	// getHeaderName returns the header name for the OAuth token.
+	getHeaderName(OAuthSpec *Middleware) string
+	//
+	headerEnabled(OAuthSpec *Middleware) bool
+}
+
+func getOAuthHeaderProvider(oauthConfig apidef.UpstreamOAuth) (OAuthHeaderProvider, error) {
+	if !oauthConfig.IsEnabled() {
+		return nil, fmt.Errorf("upstream OAuth is not enabled")
+	}
+
+	switch {
+	case len(oauthConfig.AllowedAuthorizeTypes) == 0:
+		return nil, fmt.Errorf("no OAuth configuration selected")
+	case len(oauthConfig.AllowedAuthorizeTypes) > 1:
+		return nil, fmt.Errorf("both client credentials and password authentication are provided")
+	case oauthConfig.AllowedAuthorizeTypes[0] == ClientCredentialsAuthorizeType:
+		return &ClientCredentialsOAuthProvider{}, nil
+	case oauthConfig.AllowedAuthorizeTypes[0] == PasswordAuthorizeType:
+		return &PasswordOAuthProvider{}, nil
+	default:
+		return nil, fmt.Errorf("no valid OAuth configuration provided")
+	}
+}
+
+func (p *ClientCredentialsOAuthProvider) getOAuthToken(r *http.Request, mw *Middleware) (string, error) {
+	client := UpstreamOAuthClientCredentialsClient{mw.clientCredentialsStorageHandler}
+	token, err := client.GetToken(r, mw)
+	if err != nil {
+		return handleOAuthError(r, mw, err)
+	}
+
+	return fmt.Sprintf("Bearer %s", token), nil
+}
+
+func handleOAuthError(r *http.Request, mw *Middleware, err error) (string, error) {
+	mw.EmitUpstreamOAuthEvent(r, ErrorEventName, err.Error(), mw.Spec.APIID)
+	return "", err
+}
+
+func (p *ClientCredentialsOAuthProvider) getHeaderName(OAuthSpec *Middleware) string {
+	return OAuthSpec.Spec.UpstreamAuth.OAuth.ClientCredentials.Header.Name
+}
+
+func (p *ClientCredentialsOAuthProvider) headerEnabled(OAuthSpec *Middleware) bool {
+	return OAuthSpec.Spec.UpstreamAuth.OAuth.ClientCredentials.Header.Enabled
+}
+
+type UpstreamOAuthCache interface {
```

Comment:
```suggestion
type Cache interface {
```

also move to model.go (all interfaces)